### PR TITLE
support for http verbs from RFC-2518

### PR DESCRIPTION
**Pull request by Ma27**
see https://tools.ietf.org/html/rfc2518 for further information about the http verbs

### CODE REVIEW

File: Resources/doc/5-automatic-route-generation_single-restful-controller.rst [hunk 1]
```diff
@@ -45,44 +45,62 @@ Define resource actions
...
+
+        public function proppatchUsersAction()
+        {} // "proppatch_users" [PROPPATCH] /users
+
+        public function unlockUsersAction()
+        {} // "unlock_users"    [UNLOCK] /users
+
         public function optionsUsersAction()
-        {} // "options_users" [OPTIONS] /users
+        {} // "options_users "  [OPTIONS] /users
```
Comment (Ma27) [inline Resources/doc/5-automatic-route-generation_single-restful-controller.rst]: mistyped

**Comment by Ener-Getick** [comment]
I'm not sure that we need this classes:
IMO these annotations are shortcuts and should only be available for the most common http verbs.

**Comment by Ma27** [comment]
but how common are verbs like "OPTIONS" or "LINK"?
Well love the way to configure my routes with those annotations and would love support for such http verbs, but if it you don't think that these methods should be present as annotations, feel free to close :)

**Comment by lsmith77** [comment]
I am ok with adding these annotations .. if someone needs them I guess it proofs there is a use :)

File: Routing/Loader/Reader/RestActionReader.php [hunk 2]
```diff
@@ -462,8 +480,14 @@ private function getCustomHttpMethod($httpMethod, array $resources, array $argum
...
         $annotations = array();
+        $annotationNames = array_map(
+            function ($httpVerb) {
+                return ucfirst($httpVerb);
+            },
+            $this->availableHTTPMethods
+        );
 
-        foreach (array('Route', 'Get', 'Post', 'Put', 'Patch', 'Delete', 'Link', 'Unlink', 'Head', 'Options') as $annotationName) {
+        foreach ($annotationNames as $annotationName) {
```
Comment (Ener-Getick) [inline Routing/Loader/Reader/RestActionReader.php]: I think that we can improve this:
```php
foreach($this->reader->getMethodAnnotations($reflectionMethod) as $annotation) {
    if($annotation instanceof Route) {
        $annotations[] = $annotation;
    } 
} 
```

**Comment by Ener-Getick** [comment]
Then I'm ok with it too 😉

File: Tests/Fixtures/Controller/UsersController.php [hunk 3]
```diff
@@ -58,12 +100,6 @@ public function patchUserAction($slug)
 
  // [PATCH] /users/{slug}
 
-    public function lockUserAction($slug)
```
Comment (Ener-Getick) [inline Tests/Fixtures/Controller/UsersController.php]: Why did you remove this ?

Comment (Ma27) [inline Tests/Fixtures/Controller/UsersController.php]: as the lock stuff will be handled by the lock method. The patch thing is simply a fallback: https://github.com/FriendsOfSymfony/FOSRestBundle/blob/master/Routing/Loader/Reader/RestActionReader.php#L525-L539

**Comment by Ener-Getick** [comment]
Can you fix the tests?

**Comment by Ma27** [comment]
I'm aware of some failing tests, but currently I don't have that much time, try to do this at the weekend, sry...

**Comment by Ener-Getick** [comment]
Ok no problem 😉

**Comment by JHGitty** [comment]
@Ma27 @lsmith77 
Maybe the classes should contain this comment:
```
RFC2518 - see https://tools.ietf.org/html/rfc2518
```

This would clarify the users to understand why this has been integrated. Otherwise some people could be confused.

Or we just add the information about RFC2518 just to the documentation.

**Comment by Ma27** [comment]
I'll work on this PR on the weekend if I have time to...

**Comment by Ma27** [comment]
as it contains BC breaks, I have to recreate a new PR that can be merged. Have to reopen it on 2.0 :/

**Comment by Ener-Getick** [comment]
Where do you see  bc breaks?

**Comment by Ma27** [comment]
introducing the lock method causes conflicts with the route loader.

When I wrote

``` php
public function lockUserAction() {}
```

Then the rest loader "casted" it to a PATCH method.
With the introduction of a http method called "LOCK", this behavior will be changed.

**Comment by Ener-Getick** [comment]
ah yeah ok a solution would be to support both http methods for ``LOCK``

**Comment by Ma27** [comment]
hmm, will be difficult as "PATCH" worked as a fallback.
I'm working on the tests and re-create a PR to be merged into master (I'll
use FOSRestBundle 2.0 when it comes out, so at least for me it won't be a
problem).

2016-01-15 17:31 GMT+01:00 Ener-Getick <notifications@github.com>:
